content/blog/r-mpi-part-two.md

```diff
@@ -0,0 +1,311 @@
+```
+
+and we implement it in `data/rmpi.cpp` thus :-
+
+```
+[[cpp11::register]]
+void mpi_all_reduce(cpp11::external_pointer<std::vector<double>> ptr) {
+  std::vector<double> *d = cpp11::as_cpp<cpp11::external_pointer<std::vector<double>>>(ptr).get();
+  MPI_Allreduce(MPI_IN_PLACE, d->data(), d->size(), MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+}
+```
+
+Here we are receiving the pointer that R has, to the data that's in C++ land. 
+We convert that to a pointer to actual vector, then we can pass that to 
+`MPI_Allreduce`. Normally, the first argument would be a pointer to where
+the outgoing data lives. The `MPI_IN_PLACE` flag is a special alternative
+that tells MPI the outgoing data is already in the receive buffer (the second
+argument), and to re-use that structure for receiving the result. Without that,
+we would need to double our RAM usage with an extra structure for receiving the
+result.
+
+Let's run it. After building and installing the package, I ran it
+locally with (for example) 
+`mpiexec -n 8 Rscript -e 'mpitest:::naive_reduce()'` - 
+and on our cluster using the same script from part 1, but setting 
+`/numnodes` to be the same as the number of processes - one process
+per computer. Running with between 1 and 8 processes :-
+
+![Native MPI performance](/img/mpi_2_eg1.png)
+
+Some things we can notice here :-
+
+* The time taken to do the MPI reduction (blue) increases linearly with
+  processes when run locally. We'd expect this because each process needs
+  800Mb of OS RAM. Doing the reduce as effectively a RAM operation on one
+  computer takes more time as the amount of RAM goes up. The local time (green) 
+  also suffers in the same way, because the OS takes longer to give all the
+  processes the memory they request, as the number of processes increases.
+* On the multi-node runs though, the MPI work is spread across the nodes, so
+  we don't get that degradation we got when we were running everything on one
+  node. The local time (green) also gets better as
+  we add nodes, because we really are doing less work per node. 
+* However, we don't get great _efficiency_ - doubling the number of processes
+  doesn't half the time taken, even when only considering the green line. This
+  is mainly because of [Amdahl's law](https://en.wikipedia.org/wiki/Amdahl%27s_law) - 
+  the time taken to allocate memory is included in my local time and no amount
+  of parallelisation will make that faster. Further, we're not really doing
+  that much work in the loop, so the sequential part of getting the memory
+  is a significant chunk of the total time. That limits how much speed-up
+  we could ever achieve.
+
+# Lightening the load
+
+What if we relax the problem a little. Suppose instead of all processes
+needing all of the results back, only process zero needs the assembled
+bulk. All the other processes could then only allocate memory for the 
+data they create, and contribute just that to the MPI call.
```

Comment:
Am I right that this is the thought process as we shift from shared memory to message passing as the paradigm? It might be worth linking the rust book chapter on this?